TB NLnet - project theoretical basics (#221)
3mdeb/news-and-ideas: blog/content/post/2020-03-27-trenchboot-nlnet-lz.md

---
title: Open Source DRTM with TrenchBoot. Project's basics.

> abstract: Today, we will get you into project structure and environment. You

Reviewer: We don't know when someone will read this post. I would suggest to change abstract to something like this:

```
This is the first blog post from the Trenchboot series. It will introduce you to the project, its
structure and environment. Additionally the reader will find out more about each project component,
how to setup the environment and configure the build. 
```

Author: Done

          will find out more about each component and how properly configure
          your build.

> cover: /covers/

Reviewer: Do not forget to provide a cover image

Author: Logo is added

author: piotr.kleinschmidt
layout: post
published: false
date: 2020-03-27
archives: "2020"

tags:
  - trenchboot
  - security
  - open-source
  - coreboot
categories:
  - Firmware
  - Security

---

If you haven't read [introductory blog post](link-to-introductory) yet, I
strongly recommend to do it before proceeding here. We have presented our
motivation, project's overview concept and goals there. Generally speaking,

> over time, we will be introducing new issues according to project's state. If

Reviewer: ```suggestion
we will be introducing new issues according to project's state over time. If
```

Author: Done

you want to keep up with project development, stay tuned and always try to catch
up missed articles.

> Those articles will not always be strictly technical. Therefore, if you think
some topics should be elaborated, feel free to ask us a question.

> In this article, I will explain how the project is build. Mostly, what hardware,

Reviewer: ```suggestion
In this article, I will explain how the project is built. Mostly, what hardware,
```

Author: Done

> In this article, I will explain how the project is build. Mostly, what hardware,

Author: Done

firmware and software components we are using. The most essential parts of
system will be also described in details. It's important to understand their

> role, so you will be able to, fully consciously, reproduce the environment.

Reviewer: ```suggestion
role, so you will be able to reproduce the environment smoothly and consciously.
```

Author: Done

## Testing platform

As you already know, our solution is built for AMD processors. Therefore, for
tests we use suitable platforms. Our hardware infrastructure is still being

> expanded. Eventually, there will be few different devices with different

Reviewer: ```suggestion
expanded. Eventually, there will be few more devices with different
```

> characteristic. For now, we use **PC Engines apu2** which has **AMD GX-412TC**

Author: Done

> processor. We use connection via serial port, so every relevant output will

Reviewer: ```suggestion
processor. We use connection via serial port, so every relevant output will be
```

presented in form of logs.

## Operating system

To better understand next section, first I will introduce operating system which
we use. Our choice is [NixOS](https://nixos.org/nixos/) which is a Linux

> distribution based on unusual (and reliable) package management. You can install

Author: Done

tools from official packages or build your own ones. Second feature will be

> strongly utilizes in entire project. Later, I will show you how to build custom

Reviewer: ```suggestion
strongly utilized in entire project. Later, I will show you how to build custom
```

Author: Done

> strongly utilizes in entire project. Later, I will show you how to build custom

Author: Done

package on `flashrom` example.

> Of course, you can use any other Linux distribution. However, every procedure
will be carried out by us in NixOS. If you won't have all dedicated tools or

> wrong configuration, we can't provide full reliability.

Author: @krystian-hebel I think original message reflects what I meant. No dedicated tools or wrong configuration will cause non-reliability.

> wrong configuration, we can't provide full reliability.

Reviewer 2: @piotr-kleins the way I read it is `If you won't have (...) wrong configuration...`

> wrong configuration, we can't provide full reliability.

Reviewer 2: Also, `won't` -> `don't` would be better IMO

> wrong configuration, we can't provide full reliability.

Reviewer: don't is better

#### NixOS installation

Now, I will guide through installation of NixOS. After it, you should have

> operating system in default configuration. Later, I will show you how to

Reviewer: ```suggestion
an operating system in a default configuration. Later, I will show you how to
```

Author: Done

customize it in a runtime.

>Important: NixOS installation is not interactive as in most Linux distributions
(e.g. Debian). Therefore, pay attention to typed commands and configurations!

1. Download minimal installer image from the [website](https://nixos.org/nixos/download.html).

2. Flash USB drive with the image.

> 3. Insert USB drive and choose proper boot partition.

Author: Done

Gstarting device mapper and LVM...
mounting tmpfs on /...
waiting for device /dev/root to appear...
mounting /dev/root on /iso...
mounting /mnt-root/iso/nix-store.squashfs on /nix/.ro-store...
mounting tmpfs on /nix/.rw-store...
mounting unionfs on /nix/store...

<<< NixOS Stage 2 >>>

running activation script...
setting up /etc...
unpacking the NixOS/Nixpkgs sources...
created 1 symlinks in user environment
ln: failed to create symbolic link '/root/.nix-defexpr/channels/channels': Read-only file system
starting systemd...

Welcome to NixOS 19.09.2201.7d31bbceaa1 (Loris)!

[  OK  ] Created slice system-getty.slice.
(...)
[  OK  ] Started Login Service.

<<< Welcome to NixOS 19.09.2201.7d31bbceaa1 (x86_64) - ttyS0 >>>
The "nixos" and "root" accounts have empty passwords.

Type `sudo systemctl start sshd` to start the SSH daemon.
You then must set a password for either "root" or "nixos"
with `passwd` to be able to login.

Run `nixos-help` or press <Alt-F8> for the NixOS manual.

nixos login: nixos (automatic login)
```

6. Create legacy boot partition

  >Be careful and choose correct /dev/sdX device. In our case it is `sda`, which
  is SSD disk.

  - Create a MBR partition table

>   ```bash

Author: Fixed in each occurrence.

have already installed only basic tools and applications. Additional ones can be
added using **package manager**. Most of the packages could be found in
`nixpkgs` library. All are listed in
[all-packages.nix](https://github.com/NixOS/nixpkgs/blob/master/pkgs/top-level/all-packages.nix).
Alternatively you can use the command:

```bash
nix-env -qaP '*' --description |grep -i name
```

As you can see most tools are presented in `nixpkgs`. However, there is
possibility to build your own (custom ones), based on github repository. Let's
try to this with custom `flashrom` and [3mdeb github fork](https://github.com/3mdeb/flashrom/commit/5f164cc28fdc055c272d21c60a0a32dc23d29e3b).

1. First, clone the `nixpkgs` repo. Without it, you won't have access to any
package.

```bash
git clone git@github.com:NixOS/nixpkgs.git
```

2. Change directory to `nixpkgs` and create the package directory for new
flashrom. We named it `flashrom2`, as `flashrom` already exists.

```bash
cd nixpkgs
mkdir pkgs/tools/misc/flashrom2
```

3. Move to the `flashrom2` directory and create a `default.nix` config file.

```bash
cd pkgs/tools/misc/flashrom2
touch default.nix
vim default.nix
```

> 4. Fill `default.nix` file with following content

Reviewer: I would describe what these mean. Probably these are dependencies, right?

Author: Done

```bash
{ lib
, stdenv
, fetchurl
, meson
, ninja
, pkgconfig
, libftdi1
, libusb1
, pciutils
}:

stdenv.mkDerivation rec {
  pname = "flashrom2";
  version = "2.0";

  src = builtins.fetchGit {
  url = "https://github.com/3mdeb/flashrom.git";
  ref = "wip";
  rev = "5f164cc28fdc055c272d21c60a0a32dc23d29e3b";
  };

  nativeBuildInputs = [ meson pkgconfig ninja ];
  buildInputs = [ libftdi1 libusb1 pciutils ];

  meta = with lib; {
    homepage = http://www.flashrom.org;
    description = "Utility for reading, writing, erasing and verifying flash ROM chips";
    license = licenses.gpl2;
    platforms = platforms.all;
  };
}
```

5. Build `flashrom2`

```bash
nix-build -A flashrom2
```

> `nix-build` command builds any package which is added or already exists in
`nixpkgs` library.

6. Add `flashrom2` to NixOS profile to create symlinks.

```bash
nix-env -f . -iA flashrom2
installing 'flashrom2-2.0'
building '/nix/store/hbxi8l93x2qv3kzg7kjpfaa6pmkij48f-user-environment.drv'...
created 21 symlinks in user environment
```

You have just built your first package in NixOS. Moreover, it is your custom
build! However, as you can see entire process is non-trivial. Essential part

> each pakcage is `stdenv` library, which

Reviewer 2: ```suggestion
of each package is `stdenv` library, its
```

Author: Done

[documentation](https://nixos.org/nixos/nix-pills/fundamentals-of-stdenv.html)
should clarify some issues. However, it is good to experiment with package
manager on your own. Good point to start are already existing packages and their
`default.nix` files.

You may wonder why do we build our own tools, when most of them are available in
default `nixpkgs`. Answer is simple - we have possibility to build not only
tools, but also many other software and firmware components... You will
learn more about it in next section.

## Firmware

As I mentioned in previous section, with NixOS we have possibility to customize
our platform (both firmware and operating system). Before, we do that, I need to
introduce some concepts. The better you understand them, the more awareness
about security of your platform you will have and better understand the purpose
of entire project. Let's break our entire system into following main parts. They
are listed in order of execution when the platform is booted, so you can also
see the flow and relations between them.

#### BIOS

> Our solution is fully open-source, so we also use such firmware (BIOS) -

Reviewer: ```suggestion
Our solution is open-source, so we also use such firmware (BIOS) -
```
We still have AGESA blob, unfortunately.

**coreboot**. Also, as we are maintainers of coreboot for PC Engines platforms,
we use

> [this](https://github.com/pcengines/coreboot/tree/pcengines_trenchboot_4.11.x)

Reviewer: Since we launch GRUB from disk, the branch shouldn't matter, it could be a typical release like v4.11.0.4 for example.

Author: Give reference to PC Engines/coreboot repository

particular fork and branch. Every change appears there, so it is definitely
place where you will find up-to-date firmware. I won't describe coreboot itself
here. What you need to know, it is first code which is running during boot
process. It initialize all hardware components on platform.

#### GRUB

When coreboot ends its work, a bootloader comes in. We use **GRUB**. Its task is
to boot operating system. Normally, if NixOS is installed on hard disk, GRUB
loads it to RAM. Being more precise, *Linux kernel* and *initrd* are being
loaded. However, that operation is slightly different when we enable DRTM. As it
is main goal of our project, let's take a look a little bit closer on it.

We prepared extension module called `slaunch`. When DRTM is enabled, `slaunch`

> executes AMD's dedicated machine instruction `SKINIT`. Treat is 'pre-DRTM'

Reviewer: ```suggestion
executes AMD's dedicated machine instruction `SKINIT`. Treat it as 'pre-DRTM'
```
This is what you meant?

Author: Yes, I rephrased it

> executes AMD's dedicated machine instruction `SKINIT`. Treat is 'pre-DRTM'

Reviewer 2: Please double check if I wrote it correctly
```suggestion
executes AMD's dedicated machine instruction `SKINIT`. Treat is 'DLEvent'
```

Author: Actually, I wanted to define SKINIT as 'pre-DRTM'. It is simplification, but should be suitable for non-technical users too.

operation. It prepares all components and jump to secure loader - `Landing
Zone`.

#### Landing Zone (LZ)

LZ has 2 main tasks: measure Linux kernel and then run it. If you wonder, when

> measurements are done, then you have just find out. Without LZ, DRTM doesn't

Reviewer: ```suggestion
measurements are done, then you have just found out. Without LZ, DRTM doesn't
```
Something like this?

exist. Although it seems inconspicuous, it does fundamental job in secure boot
process.

#### Linux kernel

Linux kernel is a core of all Linux based operating systems. Actually, it is
operating system without any additional applications. It can be modified to
given platform by including and excluding particular drivers, called modules. In

> general, it is done to limit size of kernel image and adjust its performance.

Reviewer: Linux kernel also does some measurements in the Trenchboot Secure Launch process.

Author: I don't mention it here, but later in 'linux-trenchboot' package description.

#### initrd

initrd stands for *initial ramdisk*. It is image used by bootloader to mount
initial file system in memory. Also, that image contains modules which will be
loaded by Linux kernel.

## Bring it all together

I have mentioned and described all parts of system which you need to know, to
understand TrenchBoot and DRTM. But how they are related and how to build such
system? I said that package manager in NixOS is very important and we will
commonly use it. And now I want to show you how.

First, we have 'clean' NixOS operating system installed on hard drive. It has

> default Linux kernel and initrd. From GRUB we can boot to it and... that's it.

Reviewer: ```suggestion
default Linux kernel and initrd. From GRUB we can boot it and... that's it.
```

Author: Done

> No DRTM, no measurements and no secure. But, with custom package management, we

Reviewer: ```suggestion
No DRTM, no measurements and no security. But, with custom package management, we
```

Author: Done

can **enable DRTM** and **prepare all components** directly from operating
system. That's what we did. We have prepared, in separate packages, all elements
mentioned above:

> - GRUB-trenchboot - it updates GRUB, so it boot `slaunch` and in consequence LZ

Reviewer: ```suggestion
- GRUB-trenchboot - it updates GRUB, so it boots `slaunch` and in consequence LZ
```

Author: Done

- LZ-trenchboot - it adds LZ which measures and runs Linux kernel (enables DRTM)
- linux-kernel-trenchboot - customized NixOS kernel
- initrd-trenchboot - customized initrd

Installing those packages will create a target system with enabled DRTM. How to
do this in practice and verify if it really works will be discussed in next

> article! So definitely, wait for next post to use this knowledge in practice.

Reviewer: It is a little bit weird to encourage somebody to do something secure, but wait a few days to know what it will do to your machine... What should we do about it? @pietrushnic cc

Author: As both articles will be published in the same day, I rephrase it to indicate to next article immediately.

strongly recommend to do it before proceeding here. We have presented our
motivation, project's overview concept and goals there. Generally speaking, we
will be introducing new issues according to project's state over time. If you
want to keep up with project development, stay tuned and always try to catch up
missed articles.

> Those articles will not always be strictly technical. Therefore, if you think
some topics should be elaborated, feel free to ask us a question.

In this article, I will explain how the project is built. Mostly, what hardware,
firmware and software components we are using. The most essential parts of
system will be also described in details. It's important to understand their
role, so you will be able to reproduce the environment smoothly and consciously.

## Testing platform

As you already know, our solution is built for AMD processors. Therefore, for
tests we use suitable platforms. Our hardware infrastructure is still being
expanded. Eventually, there will be few different devices with different
characteristic. For now, we use **PC Engines apu2** which has **AMD GX-412TC**
processor. We use connection via serial port, so every relevant output will

> presented in form of logs. What is important, our platform is equipped in

Reviewer 2: ```suggestion
presented in form of logs. What is important, our platform is equipped with
```

You may wonder why do we build our own tools, when most of them are available in
default `nixpkgs`. Answer is simple - we have possibility to build not only
tools, but also many other software and firmware components... You will
learn more about it in next section.

## Firmware

As I mentioned in previous section, with NixOS we have possibility to customize
our platform (both firmware and operating system). Before, we do that, I need to
introduce some concepts. The better you understand them, the more awareness
about security of your platform you will have and better understand the purpose
of entire project. Let's break our entire system into following main parts. They
are listed in order of execution when the platform is booted, so you can also
see the flow and relations between them.

#### BIOS

Our solution is open-source, so we also use such firmware (BIOS) - **coreboot**.
Also, as we are maintainers of coreboot for PC Engines platforms, we use
[PC Engines coreboot](https://github.com/pcengines/coreboot) and its mainline
release. Every change appears there, so it is definitely place where you will
find up-to-date firmware. I won't describe coreboot itself here. What you need

> to know, it is first code which is running during boot process. It initialize

Reviewer: ```suggestion
to know, it is first code which is running during boot process. It initializes
```

to boot operating system. Normally, if NixOS is installed on hard disk, GRUB
loads it to RAM. Being more precise, *Linux kernel* and *initrd* are being
loaded. However, that operation is slightly different when we enable DRTM. As it
is main goal of our project, let's take a look a little bit closer on it.

We prepared extension module called `slaunch`. When DRTM is enabled, `slaunch`
executes AMD's dedicated machine instruction `SKINIT`. Treat it as 'pre-DRTM'
operation. It prepares all components and jump to secure loader - `Landing
Zone`.

#### Landing Zone (LZ)

LZ has 2 main tasks: measure Linux kernel and then run it. If you wonder, when
measurements are done, then you have just found out. Without LZ, DRTM doesn't
exist. Although it seems inconspicuous, it does fundamental job in secure boot
process.

#### Linux kernel

Linux kernel is a core of all Linux based operating systems. Actually, it is
operating system without any additional applications. It can be modified to
given platform by including and excluding particular drivers, called modules. In
general, it is done to limit size of kernel image and adjust its performance. In

> Trenchboot Secure Launch process it perform additional operations - makes some

Reviewer: ```suggestion
Trenchboot Secure Launch process it performs additional operations - makes some
```

measurements. But more about it, we will discuss later.

#### initrd

> initrd stands for *initial ramdisk*. It is image used by bootloader to mount

Reviewer: ```suggestion
initrd stands for *initial ramdisk*. It is an image used by bootloader to mount
```

initial file system in memory. Also, that image contains modules which will be
loaded by Linux kernel.

## Bring it all together

I have mentioned and described all parts of system which you need to know, to
understand TrenchBoot and DRTM. But how they are related and how to build such
system? I said that package manager in NixOS is very important and we will
commonly use it. And now I want to show you how.

First, we have 'clean' NixOS operating system installed on hard drive. It has
default Linux kernel and initrd. From GRUB we can boot it and... that's it. No
DRTM, no measurements and no security. But, with custom package management, we
can **enable DRTM** and **prepare all components** directly from operating
system. That's what we did. We have prepared, in separate packages, all elements

> mentioned above:

Reviewer: ```suggestion
mentioned below:
```

- GRUB-trenchboot - it updates GRUB, so it boots `slaunch` and in consequence LZ
- LZ-trenchboot - it adds LZ which measures and runs Linux kernel (enables DRTM)

> - linux-kernel-trenchboot - customized NixOS kernel which also make some

Reviewer: ```suggestion
- linux-kernel-trenchboot - customized NixOS kernel which also makes some
```
